# AIS 01 Tutorial - Introduction

## Course organizsation

-  __Course times__
    - Lecture: Mon	12:15 - 13:45		HS Alte IHK
    - Tutorial: Wed 14:15 - 15:45		CIP II
    
    
 -  __Tutorial__
    -  Work on weekly assignments and programming examples
    -  Worksheet submissions are mandatory
    
    
 -  __WueCampus (Access Code: AIS19)__
    - Lecture slides
    - Exercises
    - Forum

## Learning Objective

- __Computational thinking__: Apply information technology, computational thinking, and utilize programming languages and software for data analysis
- __Data Management__: Data entry and annotation, data preparation and manipulation
- __Statistical techniques__: General statistical analysis techniques and their use for data inspection, exploration, and analysis
- __Machine Learning__: Effectively use approaches to perform Predictive Analytics


## Learning Environment

### Jupyter Notebooks
...

## We will be using R

R is not the only language that can be used for data analysis. Why R rather than another? Here is a list:
- interactive language
- data structures & functions
- graphics
- packages & community!

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/1/1b/R_logo.svg/724px-R_logo.svg.png" style="width:300px; float:center">

http://www.burns-stat.com/documents/tutorials/why-use-the-r-language/



### Tidyverse

The tidyverse is an opinionated collection of R packages designed for data science. All packages share an underlying design philosophy, grammar, and data structures. 

<img src="https://www.tidyverse.org/images/hex-tidyverse.png" style="width:300px">

#### Install and load packages

```R
# install
install.packages('tidyverse')
# load
library(tidyverse)
```

Tidyverse is already pre-installed, so we just need to load it:

In [1]:
library(tidyverse)

Registered S3 methods overwritten by 'ggplot2':
  method         from 
  [.quosures     rlang
  c.quosures     rlang
  print.quosures rlang
Registered S3 method overwritten by 'rvest':
  method            from
  read_xml.response xml2
── Attaching packages ─────────────────────────────────────── tidyverse 1.2.1 ──
✔ ggplot2 3.1.1       ✔ purrr   0.3.2  
✔ tibble  2.1.1       ✔ dplyr   0.8.0.1
✔ tidyr   0.8.3       ✔ stringr 1.4.0  
✔ readr   1.3.1       ✔ forcats 0.4.0  
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()


## Basic functions

### Loading packages

### Getting to know your data

Student Performance Data Set from https://rstudio-pubs-static.s3.amazonaws.com/108835_65a73467d96f4c79a5f808f5b8833922.html

In [ ]:
#### math_student_data_url <- "https://git.io/fjcqH"
math_student_data <- read.table(file = math_student_data_url, header = TRUE, sep = ";");
por_student_data_url <- "https://git.io/fjcq5";
por_student_data <- read.table(file = por_student_data_url, header = TRUE, sep = ";");

In [ ]:
math_student_data %>%
    group

## Exam Questions

### Exam AIS WS 2018/19: Data Engineering & Integration (5 points)

(a) __Analytic Pipelines__: Consider the following diamonds data set:

In [11]:
# This code was not included in the exam, values may differ
set.seed(5)
ggplot2::diamonds %>% 
    sample_n(10) %>% 
    arrange(cut) -> diamonds

In [12]:
diamonds

carat cut       color clarity depth table price x    y    z   
1  0.43  Good      F     SI1     63.6  53     948  4.83 4.79 3.06
2  0.71  Good      I     VS2     64.0  59    1840  5.61 5.55 3.57
3  0.56  Very Good H     VVS2    59.8  57    1723  5.39 5.48 3.25
4  1.13  Very Good I     SI2     59.9  57    4195  6.76 6.80 4.06
5  1.46  Premium   H     SI2     61.4  59    7604  7.30 7.26 4.47
6  1.29  Premium   J     VS1     61.7  59    5463  6.93 6.97 4.29
7  1.09  Premium   H     VS1     60.2  61    5951  6.70 6.66 4.02
8  0.53  Ideal     F     VS1     61.7  56    1630  5.20 5.22 3.21
9  0.70  Ideal     G     VS2     61.8  54    2593  5.71 5.74 3.54
10 0.32  Ideal     F     VVS2    62.0  55     786  4.39 4.41 2.73

i. (1 points) You are executing the code below. How many rows does the resulting data frame contain? Briefly explain your answer.

```R
diamonds  %>%
    group_by(cut) %>%
    summarize(median(depth))
```

__Solution__:

In [7]:
diamonds %>% 
    group_by(cut) %>%
    summarize(median(depth))

cut       median(depth)
1 Good      63.80        
2 Very Good 59.85        
3 Premium   61.40        
4 Ideal     61.80

ii. (2 points) You are executing the code below. What are the column names of the resulting data frame?

```R
diamonds %>%
    group_by(clarity , color) %>%
    filter(price > 1000) %>%
    mutate(volume = x * y * z) %>%
    summarise(x = mean(carat),y = mean(price)) %>%
    mutate(z = x * y)
```

__Solution__:

In [3]:
diamonds %>%
    group_by(clarity , color) %>%
    filter(price > 1000) %>%
    mutate(volume = x * y * z) %>%
    summarise(x = mean(carat),y = mean(price)) %>%
    mutate(z = x * y) %>%
    colnames()

[1] "clarity" "color"   "x"       "y"       "z"

iii. (2 points) Explain in pseudo code (e.g., dplyr pipelines) how to obtain the following transformed table from the given data set.

<table>
<thead>
	<tr><th scope=col>color</th><th scope=col>max_price</th><th scope=col>min_price</th></tr>
</thead>
<tbody>
	<tr><td>F</td><td>1630</td><td> 786</td></tr>
	<tr><td>G</td><td>2593</td><td>2593</td></tr>
	<tr><td>H</td><td>7604</td><td>1723</td></tr>
	<tr><td>I</td><td>4195</td><td>1840</td></tr>
	<tr><td>J</td><td>5463</td><td>5463</td></tr>
</tbody>
</table>

In [19]:
options(repr.tibble.width = 100)
diamonds %>% 
        group_by(color) %>% 
        summarise(max_price = max(price),
                  min_price = min(price))

color max_price min_price
1 F     1630       786     
2 G     2593      2593     
3 H     7604      1723     
4 I     4195      1840     
5 J     5463      5463